File: algorithms/python/373.py
import heapq

class Solution:
    def kSmallestPairs(self, nums1, nums2, k):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :type k: int
        :rtype: List[List[int]]
        """
        rls = []
        if len(nums1) <= 0 or len(nums2) <= 0:
            return rls
        pq = [(nums1[i] + nums2[0], i, 0) for i in range(len(nums1))]
        heapq.heapify(pq)
        for i in range(k):
            if len(pq) <= 0:
                break
            s, idx1, idx2 = heapq.heappop(pq)
            rls.append([nums1[idx1], nums2[idx2]])
            if idx2 + 1 < len(nums2):
                heapq.heappush(pq, (nums1[idx1] + nums2[idx2+1], idx1, idx2+1))
        return rls

File: algorithms/rust/373.rs
use std::{cmp::Reverse, collections::BinaryHeap};

impl Solution {
    #[allow(unused)]
    pub fn k_smallest_pairs(nums1: Vec<i32>, nums2: Vec<i32>, k: i32) -> Vec<Vec<i32>> {
        let mut heap = BinaryHeap::new();
        for (i, num1) in nums1.iter().enumerate() {
            heap.push((Reverse(num1 + nums2[0]), i, 0));
        }

        let mut pairs = Vec::new();
        for i in 0..k {
            match heap.pop() {
                Some((sum, idx1, idx2)) => {
                    pairs.push(vec![nums1[idx1], nums2[idx2]]);
                    if idx2 + 1 < nums2.len() {
                        heap.push((Reverse(nums1[idx1] + nums2[idx2 + 1]), idx1, idx2 + 1));
                    }
                }
                None => break,
            }
        }
        pairs
    }
}

#[allow(dead_code)]
struct Solution;

fn main() {
    println!("Hello, world!");
}
